code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/AEPPushTemplateNotificationBuilder.kt

```diff
@@ -0,0 +1,957 @@
+            )
+    }
+
+    /**
+     * Adds action details to the provided [Intent].
+     *
+     * @param intent the intent
+     * @param actionUri [String] containing the action uri
+     * @param actionId `String` containing the action ID
+     */
+    private fun addActionDetailsToIntent(
+        intent: Intent,
+        actionUri: String?,
+        actionId: String?
+    ) {
+        if (!StringUtils.isNullOrEmpty(actionUri)) {
+            intent.putExtra(PushTemplateConstants.Tracking.Keys.ACTION_URI, actionUri)
+        }
+        if (!StringUtils.isNullOrEmpty(actionId)) {
+            intent.putExtra(PushTemplateConstants.Tracking.Keys.ACTION_ID, actionId)
+        }
+    }
+
+    /**
+     * Sets the sound for the legacy style notification or notification on a device less than API 25.
+     * If a sound is received from the payload, the same is used.
+     * If a sound is not received from the payload, the default sound is used.
+     *
+     * @param context the application [Context]
+     * @param notificationBuilder the [NotificationCompat.Builder]
+     * @param customSound [String] containing the custom sound file name to load from the
+     * bundled assets
+     */
+    internal fun setSound(
+        context: Context,
+        notificationBuilder: NotificationCompat.Builder,
+        customSound: String?
+    ) {
+        if (StringUtils.isNullOrEmpty(customSound)) {
+            Log.trace(
+                PushTemplateConstants.LOG_TAG,
+                SELF_TAG,
+                "No custom sound found in the push template, using the default notification sound."
+            )
+            notificationBuilder.setSound(
+                RingtoneManager.getDefaultUri(RingtoneManager.TYPE_NOTIFICATION)
+            )
+            return
+        }
+        Log.trace(
+            PushTemplateConstants.LOG_TAG,
+            SELF_TAG,
+            "Setting sound from bundle named $customSound."
+        )
+        notificationBuilder.setSound(
+            getSoundUriForResourceName(customSound, context)
+        )
+    }
+
+    /**
+     * Sets the provided image url as the large icon for the legacy style notification. If a large icon url is received
+     * from the payload, the image is downloaded and the notification style is set to
+     * BigPictureStyle. If large icon url is not received from the payload, default style is used
+     * for the notification.
+     *
+     * @param notificationBuilder the [NotificationCompat.Builder]
+     * @param imageUrl [String] containing the image url
+     * @param title `String` containing the title
+     * @param bodyText `String` containing the body text
+     */
+    internal fun setLargeIcon(
+        notificationBuilder: NotificationCompat.Builder,
+        imageUrl: String?,
+        title: String?,
+        bodyText: String?
+    ) {
+        // Quick bail out if there is no image url
+        if (StringUtils.isNullOrEmpty(imageUrl)) return
```

Comment:
MobileCore has a method to also `setLargeIconResourceID`. Can we also check that similar to small icon if `imageUrl` is missing?

Reply:
i can add support in the future for using the large icon for the given resource id. looking at the implementation, its meant to be used for setting the resource id for the notification's primary image. I think this would be useful in displaying a local image in the case where a remote image could not be downloaded.